opcua.cpp

```diff
@@ -1503,6 +1602,253 @@ SOPC_ReturnStatus OPCUA::createS2Subscription()
 	return status;
 }
 
+/**
+ * Converts a string value to a SOPC_DataValue object based on the specified BuiltinId.
+ *
+ * @param builtinId  The OPC UA BuiltinId that determines the data type of the value.
+ * @param val        A string representation of the value to be converted.
+ * @return           Pointer to a SOPC_DataValue object on success, NULL on failure.
+ */
+SOPC_DataValue *OPCUA::toDataValue(SOPC_BuiltinId builtinId, const char *val)
+{
+	// Allocate and initialize the SOPC_DataValue structure
+	SOPC_DataValue *dv = (SOPC_DataValue *)SOPC_Calloc(1, sizeof(*dv));
+	if (dv == NULL)
+	{
+		Logger::getLogger()->error("Memory allocation failed for SOPC_DataValue.");
+		return NULL;
+	}
+	SOPC_DataValue_Initialize(dv);
+
+	dv->Value.BuiltInTypeId = builtinId;
+	dv->Value.ArrayType = SOPC_VariantArrayType_SingleValue;
+
+	int scanRes = 0;     // Holds the result of sscanf operations
+	int i8 = 0;          // Temporary variable for signed 8-bit integer
+	unsigned int u8 = 0; // Temporary variable for unsigned 8-bit integer
+
+	// Parse the value based on the specified BuiltinId
+	switch (builtinId)
+	{
+	case SOPC_Boolean_Id:
+	case SOPC_Byte_Id:
+		scanRes = sscanf(val, "%u", &u8);
+		if (scanRes != 0 && u8 <= UINT8_MAX)
+		{
+			if (builtinId == SOPC_Byte_Id)
+			{
+				dv->Value.Value.Byte = (SOPC_Byte)u8;
+			}
+			else
+			{
+				dv->Value.Value.Boolean = (SOPC_Boolean)u8;
+			}
+		}
+		else
+		{
+			scanRes = 0;
+		}
+		break;
+	case SOPC_SByte_Id:
+		scanRes = sscanf(val, "%d", &i8);
+		if (scanRes != 0 && i8 <= INT8_MAX && i8 >= INT8_MIN)
+		{
+			dv->Value.Value.Sbyte = (SOPC_SByte)i8;
+		}
+		else
+		{
+			scanRes = 0;
+		}
+		break;
+	case SOPC_Int16_Id:
+		scanRes = sscanf(val, "%" SCNi16, &dv->Value.Value.Int16);
+		break;
+	case SOPC_UInt16_Id:
+		scanRes = sscanf(val, "%" SCNu16, &dv->Value.Value.Uint16);
+		break;
+	case SOPC_Int32_Id:
+		scanRes = sscanf(val, "%" SCNi32, &dv->Value.Value.Int32);
+		break;
+	case SOPC_UInt32_Id:
+		scanRes = sscanf(val, "%" SCNu32, &dv->Value.Value.Uint32);
+		break;
+	case SOPC_Int64_Id:
+		scanRes = sscanf(val, "%" SCNi64, &dv->Value.Value.Int64);
+		break;
+	case SOPC_UInt64_Id:
+		scanRes = sscanf(val, "%" SCNu64, &dv->Value.Value.Uint64);
+		break;
+	case SOPC_Float_Id:
+		scanRes = sscanf(val, "%f", &dv->Value.Value.Floatv);
+		break;
+	case SOPC_Double_Id:
+		scanRes = sscanf(val, "%lf", &dv->Value.Value.Doublev);
+		break;
+	case SOPC_String_Id:
+		SOPC_String_Initialize(&dv->Value.Value.String);
+		if (SOPC_STATUS_OK == SOPC_String_CopyFromCString(&dv->Value.Value.String, val))
+		{
+			scanRes = 1;
+		}
+		break;
+	default:
+		Logger::getLogger()->error("Unsupported BuiltinId: %d", builtinId);
```

Comment:
Point of style: explicitly cast _builtinId_ to _int_ since _builtinId_ is an enum. Same for Line 1706 just below.

Reply:
Line 1695 is fixed. Do the same for Line 1705.